### PR TITLE
Issue #12 Testing modules solved.

### PR DESCRIPTION
**Pull request by lukaszsus**
### Overview
Issue https://github.com/NeuroSYS-pl/magda/issues/12 solved.

The issue's aim is to create a simple wrapper for [magda.module.runtime.ModuleRuntime](https://github.com/NeuroSYS-pl/magda/blob/main/magda/module/runtime.py) called `ModuleTestingWrapper`. Its interface mimics [SequentialPipeline](https://github.com/NeuroSYS-pl/magda/blob/main/magda/pipeline/sequential.py) interface, but the whole class is much simpler, for example, it does not include building [Graph](https://github.com/NeuroSYS-pl/magda/blob/83bdaf3472eba7b9be6f8000b670ab50997d5a66/magda/pipeline/graph.py) which is responsible for running modules in normal magda's pipeline. `ModuleTestingWrapper` enables simple testing of modules due to automation of calling `bootstrap` and `teardown` methods and taking care of wrapping data object into [magda.module.results.ResultSet](https://github.com/NeuroSYS-pl/magda/blob/83bdaf3472eba7b9be6f8000b670ab50997d5a66/magda/module/results.py#L23) object.

### CODE REVIEW

File: magda/testing/module_testing_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        args, kwargs = wrap_data_into_resultset(arg_names, *args, **kwargs)
+        results = await call_async_or_sync_func(self.module.run, *args, **kwargs)
+        return results
+
+    async def close(self):
+        await call_async_or_sync_func(self.module.teardown)
+
+    def _get_module_run_method_arg_names(self) -> List[str]:
+        sig = signature(self.module.run)
+        return list(sig.parameters.keys())
```
Comment (jamebs) [inline magda/testing/module_testing_wrapper.py]: `module.run` has always the same set of arguments:

```py
class MyModule(Module.Runtime):
    def run(self, data: Module.ResultSet, request: Any, *args, **kwargs):
        pass
```

Checking the method signature is not needed here

File: magda/testing/utils.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    # run does not expect 'data' argument and it has not been given
+    return args, kwargs
+
+
+async def call_async_or_sync_func(func: Callable, *args, **kwargs):
+    if asyncio.iscoroutinefunction(func):
+        return await func(*args, **kwargs)
+    else:
+        return func(*args, **kwargs)
```
Comment (jamebs) [inline magda/testing/utils.py]: Add newline at the end of file

File: magda/testing/utils.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+    if 'data' in arg_names:
+        remaining_arg_names = list(filter(lambda x: x not in kwargs.keys(), arg_names))
+        data_index = remaining_arg_names.index('data')
+        args_list = list(args)
+        args_list[data_index] = wrap_single_object_into_resultset(args_list[data_index])
+        return tuple(args_list), kwargs
+
+    # run does not expect 'data' argument and it has not been given
+    return args, kwargs
```
Comment (jamebs) [inline magda/testing/utils.py]: You can always pass `data`/`result set` to the module because the module will always be capable of handling it

File: test/test_module_testing_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        return ProcessedData(data.get(RawData)[::-1])
+
+    async def teardown(self):
+        self.is_built: bool = False
+        return super().teardown()
+
+
+@pytest.mark.parametrize('module_cls', [ModuleAsync, ModuleSync])
+@pytest.mark.asyncio
+class TestTestingModule:
```
Comment (jamebs) [inline test/test_module_testing_wrapper.py]: Rename to: `TestModuleTestingWrapper`

File: test/test_module_testing_wrapper.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    async def test_should_process_correctly(self, module_cls):
+        module = module_cls('my-module-a').set_parameters({'param': 'value'})
+        mock = await ModuleTestingWrapper(module).build()
+
+        result = await mock.run(data=RawData('xyz'), request={'query': 'query'})
```
Comment (jamebs) [inline test/test_module_testing_wrapper.py]: It should be: `... data=[RawData('xyz')] ...` (`mock.run` should accept only iterable input / list of input results)

File: test/test_module_testing_wrapper.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+        assert result == ProcessedData('zyx')
+
+    async def test_should_not_compile(self, module_cls):
+        module = module_cls('my-module-a').set_parameters({}) # Missing required params
+
+        with pytest.raises(TypeError):
+            await ModuleTestingWrapper(module).build()
+
+    async def test_should_correctyl_bootstrap(self, module_cls):
```
Comment (jamebs) [inline test/test_module_testing_wrapper.py]: typo

**Review by jamebs — CHANGES_REQUESTED** [review]

File: magda/testing/utils.py [hunk 7]
```diff
@@ -0,0 +1,22 @@
+import asyncio
+from typing import Callable
+
+from magda.module import Module
+
+
+def wrap_into_result(result, name='', src_class=None, expose=None):
```
Comment (jamebs) [inline magda/testing/utils.py]:
```suggestion
def wrap_into_result(result, name='testing-module', src_class=None, expose=None):
```

File: test/test_module_testing_wrapper.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+        await mock.close()
+        assert not mock.module.is_built
+
+    async def test_should_correctly_set_request(self, module_cls):
+        module = module_cls('my-module-a').set_parameters(TestModuleTestingWrapper.module_params)
+        mock = await ModuleTestingWrapper(module).build()
+        request = {'key': 'value'}
+        mock = mock.request(request)
+        assert isinstance(mock, ModuleTestingWrapper)
+        assert mock._request == request
```
Comment (jamebs) [inline test/test_module_testing_wrapper.py]: Instead checking the internal `_request`, I'll patch `module.run` and check if it was called with a correct `request` as a argument

File: test/test_module_testing_wrapper.py [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+
+    async def test_should_correctly_set_data(self, module_cls):
+        module = module_cls('my-module-a').set_parameters(TestModuleTestingWrapper.module_params)
+        mock = await ModuleTestingWrapper(module).build()
+        data: RawData = RawData('xyz')
+        mock = mock.data(data)
+
+        assert isinstance(mock, ModuleTestingWrapper)
+        assert isinstance(mock._data, ResultSet)
+        assert mock._data.collection[0].result.data == data.data
```
Comment (jamebs) [inline test/test_module_testing_wrapper.py]: Same here - *patch* `module.run` method and check what was passed as argument

File: test/test_module_testing_wrapper.py [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+    async def test_should_process_correctly(self):
+        module = ModuleManyDataElements('my-module-a').set_parameters({'param': 'value'})
+        mock = await ModuleTestingWrapper(module).build()
+
+        mock = mock.data(RawData('xyz'), ProcessedData('xyz'))
+        result = await mock.run()
+        assert result == ProcessedData('zyx_xyz')
+
+
+class TestWrapIntoResult:
```
Comment (jamebs) [inline test/test_module_testing_wrapper.py]: Move this class to a separate file, which will test *utils* (like `wrap_into_result`)

Comment (lukaszsus) [reply test/test_module_testing_wrapper.py]: Moved to `test/test_testing_utils.py` to distinguish `magda/testing/utils.py` from other files called `utils.py` that can appear in the future.

File: test/test_module_testing_wrapper.py (hunk 8 above)
Comment (lukaszsus) [reply test/test_module_testing_wrapper.py]: I am not fully convinced that this solution is much better. One of possible solutions requires implementing 4 'run_mock' functions:
- async and sync version for checking `request`,
- async and sync version for checking `data`.

Moreover, we are going to mock our own classes, written especially for tests. I think it will make the code little chaotic.

Comment (lukaszsus) [reply test/test_module_testing_wrapper.py]: Finally, I have implemented this with `MagickMock`. :slightly_smiling_face: Please, check if current solution fits your expectations. ;-)

File: test/test_module_testing_wrapper.py (hunk 9 above)
Comment (lukaszsus) [reply test/test_module_testing_wrapper.py]: Unfortunately, it isn't possible to implement this as I did in case of [test_should_correctly_set_request](https://github.com/NeuroSYS-pl/magda/pull/13#discussion_r650696871) because of the fact that we prepare `RawData` in the test function, but it is wrapped into ResultSet before being passed to `module.run` method. The ResultSet has not overwritten `__eq__` method so the interpreter cannot tell us if it is equal to other instance of this class.
Maybe, it is possible in some other way, but the MagickMock is the only one working, that I could find, and it is still quite elegant.

Comment (jamebs) [reply test/test_module_testing_wrapper.py]: You could assert *ResultSet[0].result* is *RawData*, because ResultSet only wraps the provided data

File: test/test_module_testing_wrapper.py [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+
+    async def test_should_correctly_teardown(self, module_cls):
+        module = module_cls('my-module-a').set_parameters(TestModuleTestingWrapper.module_params)
+        mock = await ModuleTestingWrapper(module).build()
+        await mock.close()
+        assert not mock.module.is_built
+
+    async def test_should_correctly_set_request(self, module_cls):
+        module = module_cls('my-module-a').set_parameters(TestModuleTestingWrapper.module_params)
+        module._derived_class.run = MagicMock(return_value=ProcessedData('zyx'))
```
Comment (jamebs) [inline test/test_module_testing_wrapper.py]: That's pretty ugly - you shouldn't relay on Module semi-private variable. How about:
1. `mock = await ModuleTestingWrapper(module).build()`
2. MagicMock `mock.run`
3. Check if `mock.run` was called with the same object `request`

You want to check just one option, because others (like building module) are covered by the previous tests

File: magda/testing/module_testing_wrapper.py [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+
+        return self
+
+    def request(self, request: Any) -> ModuleTestingWrapper:
+        self._request = request
+        return self
+
+    def data(self, *data) -> ModuleTestingWrapper:
+        results = list(map(wrap_into_result, data))
+        self._data = ResultSet(results)
```
Comment (jamebs) [inline magda/testing/module_testing_wrapper.py:33]: Not sure, which option will be more readable - this one or:
```py
self._data = ResultSet([wrap_into_result(r) for r in data])
```

File: test/test_testing_utils.py [hunk 13]
```diff
@@ -0,0 +1,25 @@
+import pytest
+from magda.module import Module
+from magda.testing.utils import wrap_into_result
+from test.test_module_testing_wrapper import RawData, ModuleSync
+
+
+class TestWrapIntoResult:
+    def test_should_pass(self):
```
Comment (jamebs) [inline test/test_testing_utils.py]:
```suggestion
    def test_should_wrap_data_into_result(self):
```

File: test/test_testing_utils.py [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+        result: Module.Result = wrap_into_result(data, name, src_class, expose)
+
+        assert isinstance(result, Module.Result)
+        assert result.result == data
+        assert result.interface == interface
+        assert result.name == name
+        assert result.src_class == src_class
+        assert result.expose == expose
+
+    def test_should_fail(self):
```
Comment (jamebs) [inline test/test_testing_utils.py]: Test edge cases like:
- `test_should_rise_for_missing_data`
- `test_should_accept_missing_arguments`

**Review by jamebs — CHANGES_REQUESTED** [review]
Looks fine! Just tiny 2 changes and we can merge it :smile:

**Review by jamebs — APPROVED** [review]

**Comment by jamebs** [comment]
Excellent job! Thanks for your contribution! :clap:
